### PR TITLE
ENH: making doing work better

### PR DESCRIPTION
**Pull request by joshc-slac**
The base `Worker` class was previously implemented to wrap `multiprocessing.Process` such that we could standardize process spawning and management mechanisms and propagate those changes efficiently through the codebase. 

With this in mind it made sense for our ActionNodes which inherently utilize Process's to build off of this functionality. 

## Description
We made a subclass of `Worker` called `ActionNodeWorker` which extends the Worker base class to hold the necessary interprocess communication elements needed by any ActionNode in relation to its spawned processes namely:
* __voltaile_status__ : a `multiprocessing.Value` which wraps an enum to communcate BT status up to parent node
* completion_condition: the `Callable[[None], bool]`[^1] which determines work function completion

[^1]: for full transparency this is captured by closure in the `ConditionItem` class which is partly why its valuable to pass an argument to the spawned process by the CAProcess command

This has a two fold effect:
* Ownership of IPC variables is concretized logically to the ActionNode class
* IPC nitty gritty is abstracted from the user space of defining the work function and completion condition. Currently in tree_config.py

## Motivation and Context
* With the goal of having more than "one shot" trees we need a mechanism to make a while loop that lasts the lifecycle of the program but still terminates the Worker class gives us a mechanism to do so, we may as well use it
* It is critical we standardized how work is spawned and managed lest we end up doing it differently all over the codebase which would make debugging and feature addition difficult.  

## How Has This Been Tested?
Unit tests (run better now even)

## Where Has This Been Documented?
[Slides](https://docs.google.com/presentation/d/1nmNLD5k6Zr7mg9HHmRI7F9PInfzq3mMka_9P8WJ_94E/edit?usp=sharing)

image
 

## Todo: 
In a PR based of this one we will take advantage of the "more than one shot" trees which will take full advantage of the reactivity of BTs. Two things are needed:
* make current tests more than one shot capable
* write tests that will only pass **if** they work in a more than one shot fashion. i.e. unset a PV after its been set

I am leaving these as future TODOs to keep this PR readable and make pace of dev apparent. 

## Pre-merge checklist

- [ ] Code works interactively
- [ ] Code follows the [style guide](https://pcdshub.github.io/style.html)
- [ ] Code contains descriptive docstrings, including context and API
- [ ] New/changed functions and methods are covered in the test suite where possible
- [ ] Test suite passes locally
- [ ] Test suite passes on GitHub Actions
- [ ] Ran ``docs/pre-release-notes.sh`` and created a pre-release documentation page
- [ ] Pre-release docs include context, functional descriptions, and contributors as appropriate

### CODE REVIEW

**Comment by joshc-slac** [comment]
Fixed up the merge conflicts. Waiting to push for a moment.

test_check_and_do.py had a trivial fix. 
still working on test_worker.py which has a less than trivial fix, should be resolved soon

**Comment by joshc-slac** [comment]
Closes #28

File: beams/behavior_tree/ActionNode.py [hunk 1]
```diff
@@ -84,17 +84,18 @@ def update(self) -> py_trees.common.Status:
...
-        print(f"TERMINATE CALLED ON {self.name}, pid: {os.getpid()}")
-        if self.work_proc.is_alive():
-            print(f"The process is still alive on {os.getpid()}")
-            self.work_proc.terminate()
-            self.logger.debug(
-                py_trees.console.red
-                + "%s.terminate()[%s->%s]"
-                % (self.__class__.__name__, self.status, new_status)
-            )
+    # TODO: serious introspection about that we want to do here.
```
Comment (joshc-slac) [inline beams/behavior_tree/ActionNode.py:87]: Normally anti commented code being persisted in the code base but I think this serves as an adequate TODO. Termination is a py_tree step triggered by a composite node, we likely do want to take advantage of this function not for clean up as is commented here (but maybe if we go to multi shot Processes) but for logging

File: beams/behavior_tree/ActionNode.py [hunk 2]
```diff
@@ -1,33 +1,38 @@
...
-        work_func,
-        completion_condition,
+        name: str,
+        work_func: Callable[[Any], None],
+        completion_condition: Callable[[Any], bool],
         work_gate=Event(),
         work_lock=Lock(),
         **kwargs,
     ):  # TODO: can add failure condition argument...
         super(ActionNode, self).__init__(name)
```
Comment (ZLLentz) [inline beams/behavior_tree/ActionNode.py:23]: Nitpick: this one of the old py2 syntaxes for super, it still works but it's wordier than the new way. I realize this is not part of the diff and that it matches the docs example but it caught my eye.
```suggestion
        super().__init__(name)
```

File: beams/behavior_tree/ActionNode.py [hunk 3]
```diff
@@ -84,17 +84,18 @@ def update(self) -> py_trees.common.Status:
...
+    #     """Nothing to clean up in this example."""
+    #     print(f"TERMINATE CALLED ON {self.name}, pid: {os.getpid()}")
+    #     if self.work_proc.is_alive():
+    #         print(f"The process is still alive on {os.getpid()}")
+    #         self.work_proc.terminate()
+    #         self.logger.debug(
+    #             py_trees.console.red
+    #             + "%s.terminate()[%s->%s]"
+    #             % (self.__class__.__name__, self.status, new_status)
+    #         )
```
Comment (ZLLentz) [inline beams/behavior_tree/ActionNode.py:98]: Given your above comment, maybe we can put in a basic log message instead of wiping the function entirely? I'm not sure precisely which info is useful here.
```suggestion
    def terminate(self, new_status: py_trees.common.Status) -> None:
        """Nothing to clean up.""
        logger.debug("%s terminated", self)
```

File: beams/behavior_tree/ActionWorker.py
Comment (ZLLentz) [inline beams/behavior_tree/ActionWorker.py:1]: This new file (and many of the old files) has camelcase naming which is against the pep8 recommendations
https://peps.python.org/pep-0008/#package-and-module-names

File: beams/behavior_tree/ActionWorker.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+                 work_func: Callable[[Any], None],
+                 comp_cond: Callable[[Any], bool],
+                 stop_func: Optional[Callable[[None], None]] = None):
+        super().__init__(proc_name=proc_name,
+                         stop_func=stop_func,
+                         work_func=work_func,
+                         proc_type=CAProcess,
+                         add_args=(comp_cond, volatile_status))
+        self.comp_condition = comp_cond
+        self.__volatile_status__ = volatile_status
```
Comment (ZLLentz) [inline beams/behavior_tree/ActionWorker.py:25]: What is `__volatile_status__` used for in workers? Does it do something already that I'm missing or is it planned for the future?

Why are these using the python double-underscore formatting wherever they appear?

File: beams/behavior_tree/CheckAndDo.py [hunk 5]
```diff
@@ -6,7 +6,7 @@
 
 class CheckAndDo(py_trees.composites.Selector):
     def __init__(self, name: str, check: ConditionNode, do: ActionNode) -> None:
-        super().__init__(name, memory=True)
+        super().__init__(name, memory=False)
```
Comment (ZLLentz) [inline beams/behavior_tree/CheckAndDo.py:9]: Knowledge check for me: does this make it so we always re-check the "check" side of the "check and do" on every tick even if the "do" is already running?

File: beams/sequencer/helpers/Worker.py [hunk 6]
```diff
@@ -6,42 +6,56 @@
...
-        if self.do_work.value:
+        if (self.do_work.value):
             logging.error("Already working, not starting work")
             return
         self.do_work.value = True
+        print(self.work_func)
+        # breakpoint()
         self.work_proc.start()
-        logging.info(f"Starting work on: {self.work_proc.pid}")
+        logging.debug(f"Starting work on: {self.work_proc.pid}")
```
Comment (ZLLentz) [inline beams/sequencer/helpers/Worker.py:41]: You've got some mixing of logging and printing here, and this isn't in the multiprocess context yet so it kind of seems more messy than intentional

File: beams/sequencer/helpers/Worker.py [hunk 7]
```diff
@@ -6,42 +6,56 @@
...
-class Worker:
-    def __init__(self, proc_name, stop_func=None, work_func=None):
-        self.do_work = Value("i", False)
+
+class Worker():
+    def __init__(self, 
+                 proc_name: str, 
+                 stop_func: Optional[Callable[[None], None]] = None, 
+                 work_func: Optional[Callable[[Any], None]] = None,
+                 proc_type: Union[Process, CAProcess] = Process,
```
Comment (ZLLentz) [inline beams/sequencer/helpers/Worker.py:19]: Tiny tiny type hinting nitpicks:
- CAProcess is a subclass of Process, so the Union is unnecessary for the type hint
- We pass in a type here rather than an object, so the type keyword is needed

```suggestion
                 proc_type: type[Process] = Process,
```

**Review by ZLLentz — COMMENTED** [review]
I think this is a good change, I remember most of my comments being nitpicks or questions

File: beams/tests/test_worker.py [hunk 8]
```diff
@@ -1,26 +1,33 @@
...
+                super().__init__(proc_name=proc_name,
+                                 work_func=work_func,
+                                 add_args=())
+                self.val = Value('i', 0)
 
-        assert w.value.value == 100
+        def work_func(self):
+            while self.do_work.value or self.val.value < 100:
+                if self.val.value < 100:  # Note: value captured via closure
+                    time.sleep(0.01)
```
Comment (tangkong) [inline beams/tests/test_worker.py:20]: Note from a call Josh and I sat on for a while, this sleep is necessary.  Without this, the loop runs fast enough to terminate and have trouble acquiring the lock (I think)

**Review by tangkong — COMMENTED** [review]
Some thoughts, I think this is largely good as well.  

I do think there is more to discuss w.r.t. how work is done, particularly when it comes to encapsulating the work functions.  Josh and I sat on a call and spent a good amount of time struggling to debug things, which might suggest there are better ways about it.  

The summary of my shower thought is that we could try to make this more event-driven, passing messages between sub-processes via `multiprocessing.Pipe`.  This aligns with the general guidance from python to [avoid sharing state when possible](https://docs.python.org/3/library/multiprocessing.html#synchronization-between-processes).  If we can avoid having to manage Locks entirely, perhaps our lives will be made easier in the future.  I envisioned an alternate formulation for work functions looking something like 

```python
def work_func(conn: multiprocessing.connection.Connection):
    value = 0
    do_work = True
    while do_work or value < 100:
        if conn.poll():   # returns true if there is data to receive
            # receive message and process it maybe, here I'm assuming it's just a bool
            do_work = conn.recv()
            # could also respond to requests for status here maybe?
        if value < 100:  # the rest of the work
            value += 10

    conn.send((value, status))
```

File: beams/behavior_tree/ActionWorker.py (hunk 4 above)
Comment (joshc-slac) [reply beams/behavior_tree/ActionWorker.py:25]: It does the functionality critical work of inter process communication in BTs. As Worker doesn't necessarily require IPC this ActionWorker class extends the Worker class for the subclass of things that will require IPC as they are part of trees. 

We had this conversation once before. I totally agree other mechanisms can be used to facilitate IPC. Pythonically `Queue`s and `Pipe`s are multiprocessing safe but so to is `Value`. Because we are strictly passing one of three values completely encapsulated by an enum I found Value much more elegant than needing to call send and receive on a pipe or pop from a queue. Value just wraps a portion of shared memory which seemed more elegant at this level of dev space. 

On our long call on Friday Robert and I sketched a similar idea that py_trees reach with a source of truth "database" in which all processes have access too; however, to prevent race conditions you would still have to share a Process safe lock between these workers which seems like doing the same thing as sharing a Value.

I wanted to indicate that future developers should only very intentionally mess with __volatile_status__ with the dunders but I agree now that that is the incorrect way to signal that. 

Totally open to Pipe's and Queue's but in this specific case (of just passing a wrapped int / enum) it seems more cumbersome than is worth. Again somewhat potato potato

Comment (joshc-slac) [reply beams/behavior_tree/ActionWorker.py:25]: It is important to communicate as long as these are seperate processes we need IPC safe mechanisms of passing information therefore it is _strictly_ necessary to have something of this sort. There may be an async implementation in which these processes (or then called python threads) all happen on the same stack and can share information easier via the GIL. I have reasons I can expoind upon of why I don't love this implementation but it would wokr.

File: beams/behavior_tree/CheckAndDo.py (hunk 5 above)
Comment (joshc-slac) [reply beams/behavior_tree/CheckAndDo.py:9]: 100%, I realized that error in this PR. We want the trees to err reactive. I.e. if a condition is no longer true or true we want to catch that even if executing another action. 

We should call this out more explicitly / write unit tests to show the difference. This is a great call out

File: beams/sequencer/helpers/Worker.py (hunk 7 above)
Comment (joshc-slac) [reply beams/sequencer/helpers/Worker.py:19]: Sweet, ya I  was slightly abusing the notation to show the users what types of things to pass here but from an OOP standpoint and generally you are totally correct.

File: beams/sequencer/helpers/Worker.py [hunk 9]
```diff
@@ -6,42 +6,56 @@
...
-            self.work_proc = Process(
-                target=self.work_func, name=self.proc_name, args=(self,)
-            )
+          self.work_func = work_func
+          self.work_proc = proc_type(target=self.work_func, name=self.proc_name, args=(self, *add_args,))
         self.stop_func = stop_func
 
     def start_work(self):
-        if self.do_work.value:
+        if (self.do_work.value):
```
[the remark this answers is not in the document]
Comment (joshc-slac) [reply beams/sequencer/helpers/Worker.py:34]: c habits die hard

File: beams/sequencer/helpers/Worker.py (hunk 6 above)
Comment (joshc-slac) [reply beams/sequencer/helpers/Worker.py:41]: yes good call out. when debugging I kinda spam both as they tend to print differently, not good

File: beams/behavior_tree/ActionWorker.py (hunk 4 above)
Comment (ZLLentz) [reply beams/behavior_tree/ActionWorker.py:25]: My original question is less conceptual and more that I don't understand precisely where in the code this volatile status is used.

File: beams/sequencer/helpers/Worker.py [hunk 10]
```diff
@@ -55,6 +69,4 @@ def work_func(self):
 
     def set_work_func(self, work_func):
         self.work_func = work_func
```
[the remark this answers is not in the document]
Comment (joshc-slac) [reply beams/sequencer/helpers/Worker.py:71]: Yes, especially with what you taught me about bound vs passed (?) class methods. Removing this functionality may be best, I wanted to leave it flexible till I found what seemed preferable. 

It seems we have agreed upon not using bound class members (means we have to pass self as an argument for spawned Processes) as it patters well to utilizing the base Worker class to spawn our work threads (and therefore concretize feature adds / debugging in spawned processes)

File: beams/behavior_tree/ActionWorker.py (hunk 4 above)
Comment (joshc-slac) [reply beams/behavior_tree/ActionWorker.py:25]: https://github.com/pcdshub/BEAMS/blob/af62d3325b017f358dffeba99f5f4eaff693b5ed/beams/behavior_tree/ActionNode.py#L60-L69

Critically here

Comment (joshc-slac) [reply beams/behavior_tree/ActionWorker.py:25]: Communicated here for example by work function / process:
https://github.com/pcdshub/BEAMS/blob/af62d3325b017f358dffeba99f5f4eaff693b5ed/beams/tree_config.py#L200-L204

Comment (ZLLentz) [reply beams/behavior_tree/ActionWorker.py:25]: Right, that's the volatile status of an `ActionNode`, which is different than an `ActionWorker` if I'm not misunderstanding, even if one is passed into the other. The `ActionNode` will contain an `ActionWorker` after this PR, but the `ActionWorker` is seemingly not modifying or checking the volatile status in any cases, despite getting a reference to it.

Comment (tangkong) [reply beams/behavior_tree/ActionWorker.py:25]: My point here is more an attempt to avoid sharing memory wholesale.  Why can't we refactor to an event-driven system?  While wrapping something in a `Value` seems more elegant, it just seems to be less natural in python-land, and brings with it all the locking problems we're seeing here.

Comment (joshc-slac) [reply beams/behavior_tree/ActionWorker.py:25]: So for Action Node to respond intellegenty when it gets ticked (non blocking Action Node update function). It needs some way of introspecting on the process. 

Robert and I had a discussion that instead of needing the IPC done here we could actually just have the function return a value, if the spawned function had not been joined we could consider that RUNNING.

However, to me that is just shifting the problem from communicating values via a shared peice of memory (managed by Value) to having to procedurally spawn and join Processes. This is that "do you want one Process that lives for the life cycle of the application or Processes spawned and joined ad hoc" conversation I mentioned a while ago; the reason we launch these Process's in the setup portion of the action node.

Industry (automation that is... webdev I think prefers the on demand process allocation framework) I've found generally prefers to have a fixed and finite number of threads / procs allocated that live the life cycle of the program. We can and should talk about this more at length. Both can be done with the existing Worker / ActionNode framework

File: beams/tests/test_worker.py (hunk 8 above)
Comment (joshc-slac) [reply beams/tests/test_worker.py:20]: Yes, I was somewhat surprised to find this to be the case in python. It is strictly necessary in a language like C in which if you were to go full bore in a loop acquiring a Lock, there would not be CPU cycles for another, lower or equiv priority thread to acquire the same lock

File: beams/behavior_tree/ActionWorker.py (hunk 4 above)
Comment (tangkong) [reply beams/behavior_tree/ActionWorker.py:25]: Definitely a further discussion point 👍 

For what it's worth, I don't think that being event-driven necessarily conflicts with the fixed-process-number goal.

File: beams/behavior_tree/ActionWorker.py
Comment (joshc-slac) [reply beams/behavior_tree/ActionWorker.py:1]: great call out, I might make this a seprate trivial PR

File: beams/behavior_tree/ActionWorker.py (hunk 4 above)
Comment (joshc-slac) [reply beams/behavior_tree/ActionWorker.py:25]: Sorry did not see yalls messages between the two. 

@ZLLentz ActionWorker needs a reference to the piece of shared memory so that Worker which spawns the Process can pass that piece of shared memory as an argument. See the add_args parameter:
https://github.com/pcdshub/BEAMS/blob/1b5910edb30a7f65eba67c2a5c87f113dcb4a8b3/beams/sequencer/helpers/Worker.py#L18

Passed to the Process here: 
https://github.com/pcdshub/BEAMS/blob/1b5910edb30a7f65eba67c2a5c87f113dcb4a8b3/beams/sequencer/helpers/Worker.py#L28

This is all to enable my not so well stated dream of abstracting this semi complex IPC stuff from 95% of developers and wanting ActionNode to own all IPC related variables:
https://github.com/pcdshub/BEAMS/blob/1b5910edb30a7f65eba67c2a5c87f113dcb4a8b3/beams/behavior_tree/ActionNode.py#L26-L33

For 90+% of users this would just mean that to package something in an ActionNode they would have to write a single function.



@tangkong I'm not sure if event driven is necessarily the term to use here. Fundamentally the ticking of these trees is a polling operation. We could make an event driven schema (in which signals determine which functions get invoked) but that would require a replumbing of py_trees itself; unless we are speaking about different things

**Comment by joshc-slac** [comment]
> Some thoughts, I think this is largely good as well.
> 
> I do think there is more to discuss w.r.t. how work is done, particularly when it comes to encapsulating the work functions. Josh and I sat on a call and spent a good amount of time struggling to debug things, which might suggest there are better ways about it.
> 
> The summary of my shower thought is that we could try to make this more event-driven, passing messages between sub-processes via `multiprocessing.Pipe`. This aligns with the general guidance from python to [avoid sharing state when possible](https://docs.python.org/3/library/multiprocessing.html#synchronization-between-processes). If we can avoid having to manage Locks entirely, perhaps our lives will be made easier in the future. I envisioned an alternate formulation for work functions looking something like
> 
> ```python
> def work_func(conn: multiprocessing.connection.Connection):
>     value = 0
>     do_work = True
>     while do_work or value < 100:
>         if conn.poll():   # returns true if there is data to receive
>             # receive message and process it maybe, here I'm assuming it's just a bool
>             do_work = conn.recv()
>             # could also respond to requests for status here maybe?
>         if value < 100:  # the rest of the work
>             value += 10
> 
>     conn.send((value, status))
> ```
> 
> maybe I'm just too event-pilled for my own good

On one hand you're right it took both of us a few hours to iron out. To be fair the largest difficulty I had was understanding the notion of bound class members vs passed class members. The locking stuff became fairly trivial after that, especially after learning python doesnt like full bore acquisition of locks. I am not sure I agree with what this work function achieves but I see the general intent. We would have to send the "volatile status" on the socket and the update method of the action node would have to recv it. There are some nuances there where I would use an Event() variable to signal that data is available on the socket to make sure we ack every write (i.e. are not checking stale data on either end) and more critically do not block in the update method. 

I do not think this maps to my notion of event driven. In which we bind functions to signals handled by an event queue. I hope I am not coming across as overly pedantic or portraying this as overly complicated these are bog standard trade offs in systems design; the hard thing is getting it right for the right reasons then it becomes a game of abstraction.

File: beams/behavior_tree/ActionWorker.py (hunk 4 above)
Comment (ZLLentz) [reply beams/behavior_tree/ActionWorker.py:25]: I'm so sorry to belabor this tiny point when the other things are more interesting, but, are you saying that the worker needs to maintain a direct reference to the volatile status after passing it into the work proc? Naively I'd assume that passing it into the super call is the last time it is needed. My initial confusion came from seeing the `self.__volatile_status__ = <thing>` only to have it never accessed that way ever again, I will admit I missed some of the other codepaths and I understand the overall data sharing flow better now.

I'm very interested in a less asynchronous discussion about any long-term goals wrt event-driven, etc. but I think the use of these memory-sharing objects is appropriate and helpful right now.

Comment (joshc-slac) [reply beams/behavior_tree/ActionWorker.py:25]: Don't be sorry its a good question. Worker needs to pass it to the constructor of Process. Worker owns the process; ActionNode which instantiates Worker by way of ActionWorker owns the IPC related variables namely here volatile status.

**Comment by joshc-slac** [comment]
A lot of good discussion. At the last beams meeting we determined the implementation details decided in this PR are sufficient to carry us on the order of many months and should this application experience more success, development hours we can revisit some of the decisions made here.

Attached are [slides](https://docs.google.com/presentation/d/1nmNLD5k6Zr7mg9HHmRI7F9PInfzq3mMka_9P8WJ_94E/edit?usp=sharing) generated to reflect the various trade offs and calculations made to settle on these implementation details.

File: beams/behavior_tree/ActionWorker.py (hunk 4 above)
Comment (ZLLentz) [reply beams/behavior_tree/ActionWorker.py:25]: I'm going to hit approve but I maintain that everything still works if you remove this specific line of code since it will never be accessed this way again during execution

**Review by ZLLentz — DISMISSED** [review]

File: beams/behavior_tree/ActionWorker.py (hunk 4 above)
Comment (joshc-slac) [reply beams/behavior_tree/ActionWorker.py:25]: Ah.... Sorry it took me till now to see your exact point. Its specifically the the ActionWorker does not need to maintain a reference to this "piece of shared memory" 

From one perspective that is **totally, totally** correct. Lets not leave a loaded gun laying around (sorry the metaphor was too fun to avoid). However, there is also a world where ActionWorker defines stop_work functions for all work that gets packaged and sent off to Worker and it may be useful to have a reference to it here... Though an argument could be made that that should be grabbed by invoking `super` in some way I am not pythonically familiar with yet...

TLDR: you're right I can remove.

Comment (tangkong) [reply beams/behavior_tree/ActionWorker.py]: There's a guideline I read somewhere that PR's should never include pieces that "will be used later on"

I catch myself doing that all too often.  It's definitely good to help keep diffs understandable

Comment (joshc-slac) [reply beams/behavior_tree/ActionWorker.py]: yeah agreed, i've had peers willing to die on the hill of commented out code should't live in commits either

**Review by ZLLentz — APPROVED** [review]

**Review by tangkong — APPROVED** [review]
On to the next
